arch/x86/deflate_quick.c

```diff
@@ -201,16 +201,35 @@ static void static_emit_end_block(deflate_state *const s, int last) {
 static inline Pos quick_insert_string(deflate_state *const s, const Pos str) {
     Pos ret;
     unsigned h = 0;
+    unsigned int val;
 
-#ifdef _MSC_VER
-    h = _mm_crc32_u32(h, *(unsigned *)(s->window + str));
+#if defined(UNALIGNED_OK)
+    val = *(unsigned int *)(s->window + str);
+#else
+    memcpy(&val, s->window + str, sizeof(val)); 
+#endif
+
+#if defined(X86_SSE42_CRC_INTRIN)
+#  ifdef _MSC_VER
+    h = _mm_crc32_u32(h, val);
+#  else
+    h = __builtin_ia32_crc32si(h, val);
+#  endif
 #else
+#  ifdef _MSC_VER
+    __asm {
+        mov edx, h
+        mov eax, val
+        crc32 eax, edx
+        mov val, eax
+    };
+#  else
     __asm__ __volatile__ (
-        "crc32l (%[window], %[str], 1), %0\n\t"
-    : "+r" (h)
-    : [window] "r" (s->window),
-      [str] "r" ((uintptr_t)str)
+        "crc32 %1,%0\n\t"
+        : "+r" (h)
+        : "r" (val)
     );
+#  endif
 #endif
```

Comment:
Instead of duplicating the code, it should be moved to private header as inline function.

Reply:
Any thoughts on the name of the function?

Reply:
Obviously it should not conflict with existing intrinsic names... and the name should clearly show the size of the data parameter (uint32)...

Reply:
It doesn't seem to be possible because of the statement:
```
        if (s->level >= TRIGGER_LEVEL)
            val &= 0xFFFFFF;
```
I wonder if that is even necessary.

Reply:
That statement is no issue as the inline function would start right after it.

Reply:
What do you think about changing insert_string to be a function in deflate.c and doing a  `functable.compute_hash` function instead? The majority of the insert_string functions are the same except for the assembly for the hash.